### DISCUSSION TITLE
Android Port

### DISCUSSION THREAD

**Discussion by hipersayanX** [webcamoid/webcamoid#519] [Ideas]
**Stage 1**: Make it work
Ready :heavy_check_mark:

![Webcamoid in Android](https://user-images.githubusercontent.com/1249885/54470132-9228a000-4781-11e9-9a3e-2fe784f9d648.png)

 :sunglasses:

**Comment by hipersayanX** [comment]
**Stage 2.0**: Camera working (both JNI and NDK) :heavy_check_mark: 

![android-camera](https://user-images.githubusercontent.com/1249885/56699657-ce2e1980-66cc-11e9-899f-c84ce1e11500.png)

**Comment by hipersayanX** [comment]
**Stage 2.1**: Screen capture working :heavy_check_mark:

![Screenshot_20190427_120914](https://user-images.githubusercontent.com/1249885/56851494-3f94e480-68e6-11e9-82d8-aeca292ecc18.png)

**Comment by hipersayanX** [comment]
**Stage 2.2**: Video and audio playback working :heavy_check_mark:

![Screenshot_20190718_172141](https://user-images.githubusercontent.com/1249885/61493701-9b718580-a98a-11e9-9c3f-66d081736e44.png)

**Comment by hipersayanX** [comment]
**Stage 2.3**: Video and audio recording working :heavy_check_mark:

![Screenshot_20190817_121437](https://user-images.githubusercontent.com/1249885/63214098-5f952180-c0ea-11e9-875c-3cf1bf62526f.png)

And now Webcamoid has all base features working in Android, obviously there would be some bugs out there that needs to be fixed, and Webcamoid is not usable at all yet, but over all the most important part of the port is done.
The next step is reworking the UI, and I'm planning a big full rewrite of it, I will keep you updated.

**Comment by hipersayanX** [comment]
[Android packages](https://bintray.com/webcamoid/webcamoid/webcamoid/daily#files/android) are ready for testing. The packages are signed with a debug sign so you will need to enable Android to install from unknown sources. You need Android 6.0 or higher.
This is **NOT** an official release, and as said before this is not ready for daily usage, expect a **LOT** of bugs.

**Comment by blackchineykh** [comment]
Hi, are these packages available for download? The link doesn't give us a way to download.

Also does this allow override of the actual physical camera on android? This is my use case and would love to see if I could use for this and also lend any support

**Comment by hipersayanX** [comment]
@blackchineykh yes in the **Files** section in the **android** folder. Bintray seems to be a bit bugged and it does not show it directly :confused:

**Comment by hipersayanX** [comment]
Ok, after 3 years of not updating, and having several problems with Android packages in CI environments, the Webcamoid port is up and running once again.

![Screenshot_20220114_192902](https://user-images.githubusercontent.com/1249885/149594854-dccab785-d243-41fc-ad6b-85d5c8bfb802.png)

The UI looks clunky, and there are some bugs to solve yet, but the most important part is that at least it's working and the port can keep going forward.

**Comment by hipersayanX** [comment]
I found that is crashing when making a screen capture, the port will need a bit more of work.

**Comment by beedell-roke** [comment]
@hipersayanX, why discussion rather an issue?

**Comment by hipersayanX** [comment]
@BEEDELLROKEJULIANLOCKHART I thought it would be better to move these kind of topics here for giving ideas, announcements, progress reports, and so on, and leave the issues page just for users reporting bugs.

**Comment by hipersayanX** [comment]
I've been testing Webcamoid the last weeks with many different configurations: phone, tablets, and the emulator, and considering that the Android port is in a development stage between alpha and beta, it's almost usable, and now is working a lot much better compared to 1 year ago.

![Screenshot_20230210_151744](https://user-images.githubusercontent.com/1249885/218169649-c9296289-98e7-4fee-9e2d-b547e1004c67.png)

# Things that have been improved

- UI scaling. As said before, I been testing Webcamoid with different screens and resolutions, and all the UI controls are scaling perfectly fine for all of them, and with small screens is usable at 90%, and the 10% missing can be safely ignored. Maybe some elements should be moved here or there for a better user experience, but beside all the basic functionality is there.
- The picture taken by the camera shows with the right orientation.
-  All the features features and bug fix added for all other platforms are now working in Android too.

# Things still failing or missing

- I was working to add flash light support, but Webcamoid was not able to detect it, maybe it needs some minor fix in the code.
- Audio playback sounds crackling, I think this is caused by the high latency between [AudioTrack.write()](https://developer.android.com/reference/android/media/AudioTrack#write(byte[],%20int,%20int,%20int)) and [AMediaExtractor](https://developer.android.com/ndk/reference/group/media#group___media_1ga2dc9581afed2b5fbf96acc58187546f0), I need to investigate more.
- Low frame rate capture, this could be because I've tested everything with debug binaries, anyway my idea is adding support for [intrinsics](https://learn.microsoft.com/en-us/cpp/intrinsics/compiler-intrinsics) (in all platforms), and move all the video processing to [Vulkan](https://www.vulkan.org/).
- The about dialog is absolutely unreadable in small screens, will have to figure out a different design.
- Some parts of the UI doesn't scroll, for example the effect list and the plugins list, this only happens in Android.
- Android 30 and higher needs some [special permissions and a convoluted access for reading multimedia files](https://developer.android.com/training/data-storage/shared/media), I didn't studied it in depth yet.

My idea is to fix the audio bugs, UI scrolling, and the multimedia playback, and make the official Android release, and fix the others bugs later.
But overall, even with the current state, I'm happy with the result :smile: 

And at last but not least, a question for you who's reading :point_up_2: 

>What would happens if I add video layering, transitions, and streaming support?

**Reply by beedell-roke** [reply]
If you added streaming support, you'd basically create the only OBS competitor for Android. What's video layering – adding multiple sources onto a canvas, like OBS does?

**Comment by hipersayanX** [comment]
@rokejulianlockhart 

>If you added streaming support, you'd basically create the only OBS competitor for Android.

Exactly :wink:  you could have a whole streaming studio in your phone without requiring a PC.

>What's video layering – adding multiple sources onto a canvas, like OBS does?

That's the idea, for now Webcamoid can show one source at time, adding layers it would be possible to for example create a scene with the screen as background and the camera on top for reactions, the same way as OBS. Also that's why I want to move all video processing to Vulkan.
Then,

>if OBS can already do that what would be the difference with Webcamoid?

Webcamoid already works in Android, while OBS needs a gigantic code redesign to port it to Android.

**Reply by beedell-roke** [reply]
Absolutely superb. That last point is indeed the reason why I have both installed. Although OBS is ultimately undoubtedly more powerful, for the features that I use, `webcamoid` already provides all of them in a consistent, platform-agnostic manner.

**Comment by hipersayanX** [comment]
Some updates:

# Fixed things

- The video playback bugs were fixed and videos now play smoothly.
- Almost all UI usability problems were fixed. The **Plugins** section in the **Preferences** menu doesn't scrolls, but this is something that most users won't need to touch, everything else is ok.

# Things still failing or missing

- I will leave the flashback support for later.
- I will try to improve frame rate capture later.
- I still need to test audio recording.

Basically the last things I need check are: audio recording (for all platforms), review the issue list for any issue with a quick or must fix, do a final check of everything, and Webcamoid will be ready for a final release.
